book/data_viz/plotting_data.nim

```diff
@@ -7,18 +7,521 @@ nbUseNimibook
+will use in all following snippets. It will contain 3 columns:
+- "Energy": a column of twice 25 elements covering the range (0, 10) with 25 entries
+- "Counts": a column of twice 25 random entries between 0 and 10. The first 25 elements
+  are drawn as floats (to get fractional values) and the second 25 entries will be random
+  integer numbers
+- "Type": simply a column that designates the first 25 rows as "background" and the latter
+  25 as "calibration"
+
+Our construction in the following is a bit artificial of course.
+"""
+nbCode:
+  import ggplotnim, random, sequtils
+  randomize(42)
+  let df = seqsToDf({ "Energy" : cycle(linspace(0.0, 10.0, 25).toRawSeq, 2),
+                      "Counts" : concat(toSeq(0 ..< 25).mapIt(rand(10.0)),
+                                        toSeq(0 ..< 25).mapIt(rand(10).float)),
+                      "Type" : concat(newSeqWith(25, "background"),
+                                      newSeqWith(25, "candidates")) })
+  echo "Input data frame: "
+  echo "Head(10): ", df.head(10).pretty(10)
+  echo "Tail(10): ", df.tail(10).pretty(10)
 nbText: """
+What we want to achieve as a final plot is the following, where the explanations
+are mainly due to the original motivation of where the plot example is taken from:
+- a histogram for each "Type", drawn with one color each and a bit transparent so they
+  are visible where they overlap. Let each bin content correspond to some data point
+  at that energy.
+- also plot the actual data points on top of the bins. For the "background" like data
+  we have fractional values as it's values are normalized to the "candidates" (simply
+  by scaling from some hypothetical time for background / time for candidate data).
+- For the background data we want error bars. They represent our uncertainty of the
+  background hypothesis.
+- the candidates are just counts we measured. They don't have inherent uncertainties
+  (from a frequentist perspective we have to repeat the experiment many times and *then*
+  we can write down some variance in our candidates)
+
+##### Building layer 1 - `geom_histogram`
+
+So, let's start with drawing a histogram of "Energy" and "Counts":
+"""
+nbCodeInBlock:
+  ggplot(df, aes("Energy", "Counts")) +
+    geom_histogram() +
+    ggsave("images/multi_layer_histogram_0.png")
+nbImage("images/multi_layer_histogram_0.png")
+nbText: """
+So, uhm. This looks rather broken! Or at least not what we want. What's going on?
+We're asking for a histogram! By default this means `ggplotnim` will *compute* the
+histogram based on the `x` aesthetic. (Note: it should at least print a warning
+message if a `y` aesthetic is given that user probably wants identity stats!).
+Instead our data is *already* binned. We need the `stat = "identity"` option to
+the `geom_histogram` call:
+"""
+nbCodeInBlock:
+  ggplot(df, aes("Energy", "Counts")) +
+    geom_histogram(stat = "identity", hdKind = hdBars) +
```

Comment:
Oops, that was me checking something. I accidentally staged that. Will be removed, thanks!

`hdKind` is an enum that can be handed to `geom_histogram` to decide the drawing style. By default (`hdBars`) a histogram is drawn as a set of individual bars that touch. That leads to ugly Moiré / aliasing effects in plots depending on zoom level where you can often see the background between bars (which shouldn't happen for touching bars of course). 

So in addition there's `hdOutline` that draws a histogram using *the outline* and then fills the whole thing with one color. Gets rid of the above problem and is more efficient (only a single polygon instead of N for N bins). 

It's not the default yet for a few reasons, mainly that it's still new and has a few problems (the last ghost bin in the final plot for instance).